api/v1alpha1/virtualcluster_types.go

```diff
@@ -0,0 +1,262 @@
+	// object. See [ClusterSpec] for more information.
+	// +kubebuilder:validation:Optional
+	Cluster *ClusterSpec `json:"cluster,omitempty"`
+
+	// Helm allows configuring helm chart options of both agent and vcluster.
+	// It is then deployed by the [VirtualCluster] CRD controller.
+	// +kubebuilder:validation:Optional
+	Helm *HelmSpec `json:"helm,omitempty"`
+}
+
+type HelmSpec struct {
+	// Agent allows configuring agent specific helm chart options.
+	// +kubebuilder:validation:Optional
+	Agent *AgentHelmConfiguration `json:"agent,omitempty"`
+
+	// VCluster allows configuring vcluster specific helm chart options.
+	// +kubebuilder:validation:Optional
+	VCluster *VClusterHelmConfiguration `json:"vcluster,omitempty"`
+}
+
+func (in *HelmSpec) GetAgent() *AgentHelmConfiguration {
+	if in == nil || in.Agent == nil {
+		return &AgentHelmConfiguration{
+			HelmConfiguration: HelmConfiguration{
+				RepoUrl:   lo.ToPtr(AgentDefaultRepository),
+				ChartName: lo.ToPtr(AgentDefaultChartName),
+			},
+		}
+	}
+
+	return in.Agent
+}
+
+func (in *HelmSpec) GetVCluster() *VClusterHelmConfiguration {
+	if in == nil || in.VCluster == nil {
+		return &VClusterHelmConfiguration{
+			HelmConfiguration: HelmConfiguration{
+				RepoUrl:   lo.ToPtr(VClusterDefaultRepository),
+				ChartName: lo.ToPtr(VClusterDefaultChartName),
+			},
+		}
+	}
+
+	return in.VCluster
+}
+
+type HelmConfiguration struct {
+	// ChartName is a helm chart name.
+	ChartName *string `json:"chartName,omitempty"`
+
+	// RepoUrl is a url that points to this helm chart.
+	// +kubebuilder:validation:Type:=string
+	// +kubebuilder:validation:Optional
+	RepoUrl *string `json:"repoUrl,omitempty"`
+
+	// Values allows defining arbitrary YAML values to pass to the helm as values.yaml file.
+	// Use only one of:
+	// 	- Values
+	//	- ValuesSecretRef
+	//	- ValuesConfigMapRef
+	// +kubebuilder:validation:Optional
+	Values *runtime.RawExtension `json:"values,omitempty"`
+
+	// ValuesSecretRef fetches helm values from a secret in this cluster.
+	// Use only one of:
+	// 	- Values
+	//	- ValuesSecretRef
+	//	- ValuesConfigMapRef
+	// +kubebuilder:validation:Optional
+	ValuesSecretRef *corev1.SecretKeySelector `json:"valuesSecretRef,omitempty"`
+
+	// ValuesConfigMapRef fetches helm values from a config map in this cluster.
+	// Use only one of:
+	// 	- Values
+	//	- ValuesSecretRef
+	//	- ValuesConfigMapRef
+	// +kubebuilder:validation:Optional
+	ValuesConfigMapRef *corev1.ConfigMapKeySelector `json:"valuesConfigMapRef,omitempty"`
+}
+
+type AgentHelmConfiguration struct {
+	HelmConfiguration `json:",inline"`
+}
+
+func (in *AgentHelmConfiguration) GetChartName() string {
+	if in == nil || in.ChartName == nil {
+		return AgentDefaultChartName
+	}
+
+	return *in.ChartName
+}
+
+func (in *AgentHelmConfiguration) GetRepoUrl() string {
+	if in == nil || in.RepoUrl == nil {
+		return AgentDefaultRepository
+	}
+
+	return *in.RepoUrl
+}
+
+type VClusterHelmConfiguration struct {
```

Comment:
Is there a reason why you inline the helm configuration struct?  Suppose it's nice for the types to self-document

Reply:
That's just not to add an additional level to access the structure. Initially I have used HelmConfiguration directly, but I wanted to have some methods that can default values required to initialize cluster/agent such as repo url or chart name. This way structure access looks the same from the outside but there are different method implementations.